torchscript_export.py

```diff
@@ -0,0 +1,126 @@
+import os
+import sys
+import argparse
+from pathlib import Path
+
+import torch
+
+# Add 'core', which contains the imports
+# for MonSter to the module search path.
+sys.path.append("core")
+
+from monster import Monster
+from utils.utils import InputPadder
+
+
+class ScriptableMonSter(torch.nn.Module):
+    def __init__(self, monster: torch.nn.Module):
+        super().__init__()
+        self.monster = monster
+
+    def forward(
+        self,
+        left_image: torch.Tensor,
+        right_image: torch.Tensor
+    ) -> torch.Tensor:
+
+        left_image = left_image.unsqueeze(0)
+        right_image = right_image.unsqueeze(0)
+
+        # Now pad to make divisible by 32
+        padder = InputPadder(left_image.shape, divis_by=32)
+        left_image, right_image = padder.pad(inputs=[left_image, right_image])
+
+        # Run the monster model
+        disp = self.monster(left_image, right_image, iters=24, test_mode=True)
+
+        # Remove padding
+        disp = padder.unpad(disp)
+
+        disp = torch.squeeze(disp[:, 0, :, :])
+
+        return disp
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--restore_ckpt",
+        help="restore checkpoint",
+        type=Path,
+        default=Path(".") / "pretrained" / "mix_all.pth",
+    )
+    parser.add_argument(
+        "--validate",
+        action="store_true",
+        help="detect offending (TorchScript incompatible) graph nodes",
+    )
+    parser.add_argument(
+        "--output_directory",
+        help="directory to save output",
+        type=Path,
+        default=Path(".") / "output",
+    )
+
+    return parser.parse_args()
+
+
+def main() -> None:
+    cli_args = parse_args()
+
+    # Load the model.
+    model = Monster()
+
+    assert os.path.exists(cli_args.restore_ckpt)
+    checkpoint = torch.load(
+        cli_args.restore_ckpt, map_location="cpu", weights_only=True
+    )
+
+    model.load_state_dict(checkpoint, strict=True)
+
+    # Wrap the model in a scriptable MonSter model.
+    monster = ScriptableMonSter(model)
+    monster = monster.eval()
+
+    # Validate whether the model is compatible for TorchScript export.
+    if cli_args.validate:
```

Comment:
What's the point of this?
If JIT compilation fails it should already throw an error indicating where it failed and why.

Reply:
The torchscript exporter will exit at the first point of failure, even if there are other sub-modules that cannot be exported. Validate runs the exporter on all the sub-modules of MonSter and lists all the _first_ points of failure in all of them.

Does it have to be run only once? No, because there could be other points of failure in the sub-modules. But this approach helped me figure out multiple problems in a single run. Otherwise we have to run the export script after every single fix that makes the model torchscript compatible.

I can improve the documentation. Let me know if this is unnecessary and we can just list the errors one by one.

Reply:
I understand, but now that it is fixed this is kinda pointless, I guess. Even if there are further incompatibilities, due to changes of the torch version, it probably won't be that many that warrants using this method, imo.